File: server/app/services/watchlist_service.py
from app import db
from app.models.watchlist import Watchlist
from app.models.asset import Asset
from app.models.portfolio import Portfolio

from datetime import datetime, timezone

def add_to_watchlist(portfolio_id, asset_id, notes=None):
    """
    Adds an asset to a portfolio's watchlist.
    """
    # Check if portfolio exists
    portfolio = Portfolio.query.get(portfolio_id)
    if not portfolio:
        raise ValueError("Portfolio not found.")

    # Check if asset exists
    asset = Asset.query.get(asset_id)
    if not asset:
        raise ValueError("Asset not found.")

    # Check if already in watchlist
    existing = Watchlist.query.filter_by(portfolio_id=portfolio_id, asset_id=asset_id).first()
    if existing:
        raise ValueError("Asset is already in watchlist.")

    # Create new watchlist entry
    watchlist_item = Watchlist(
        portfolio_id=portfolio_id,
        asset_id=asset_id,
        notes=notes
    )
    
    db.session.add(watchlist_item)
    db.session.commit()

    return watchlist_item

def remove_from_watchlist(watchlist_id):
    """
    Removes an asset from watchlist by watchlist ID.
    """
    watchlist_item = Watchlist.query.get(watchlist_id)
    if not watchlist_item:
        raise ValueError("Watchlist item not found.")

    db.session.delete(watchlist_item)
    db.session.commit()

    return True

def remove_from_watchlist_by_asset(portfolio_id, asset_id):
    """
    Removes an asset from watchlist by portfolio_id and asset_id.
    """
    watchlist_item = Watchlist.query.filter_by(portfolio_id=portfolio_id, asset_id=asset_id).first()
    if not watchlist_item:
        raise ValueError("Asset not found in watchlist.")

    db.session.delete(watchlist_item)
    db.session.commit()

    return True

def get_watchlist_by_portfolio(portfolio_id):
    """
    Gets all watchlist items for a specific portfolio.
    """
    # Check if portfolio exists
    portfolio = Portfolio.query.get(portfolio_id)
    if not portfolio:
        raise ValueError("Portfolio not found.")

    watchlist_items = Watchlist.query.filter_by(portfolio_id=portfolio_id).all()
    return watchlist_items

File: client/src/services/watchlistService.ts
import { WatchlistItem, CreateWatchlistRequest, UpdateWatchlistRequest } from '../types';
import { apiClient } from './apiClient';

export class WatchlistService {
  private static baseUrl = '/watchlist';

  // Get all watchlist items for a portfolio
  static async getWatchlistByPortfolio(portfolioId: number): Promise<WatchlistItem[]> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/portfolio/${portfolioId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching watchlist:', error);
      throw error;
    }
  }

  // Add an asset to watchlist
  static async addToWatchlist(portfolioId: number, request: CreateWatchlistRequest): Promise<WatchlistItem> {
    try {
      const response = await apiClient.post(`${this.baseUrl}/portfolio/${portfolioId}`, request);
      return response.data;
    } catch (error) {
      console.error('Error adding to watchlist:', error);
      throw error;
    }
  }

  // Remove an asset from watchlist by watchlist item ID
  static async removeFromWatchlist(watchlistId: number): Promise<void> {
    try {
      await apiClient.delete(`${this.baseUrl}/${watchlistId}`);
    } catch (error) {
      console.error('Error removing from watchlist:', error);
      throw error;
    }
  }

  // Remove an asset from watchlist by portfolio ID and asset ID
  static async removeFromWatchlistByAsset(portfolioId: number, assetId: number): Promise<void> {
    try {
      await apiClient.delete(`${this.baseUrl}/portfolio/${portfolioId}/asset/${assetId}`);
    } catch (error) {
      console.error('Error removing asset from watchlist:', error);
      throw error;
    }
  }

  // Update watchlist item notes
  static async updateWatchlistNotes(watchlistId: number, request: UpdateWatchlistRequest): Promise<WatchlistItem> {
    try {
      const response = await apiClient.put(`${this.baseUrl}/${watchlistId}`, request);
      return response.data;
    } catch (error) {
      console.error('Error updating watchlist notes:', error);
      throw error;
    }
  }

  // Get a specific watchlist item
  static async getWatchlistItem(watchlistId: number): Promise<WatchlistItem> {
    try {
      const response = await apiClient.get(`${this.baseUrl}/${watchlistId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching watchlist item:', error);
      throw error;
    }
  }

  // Get all watchlist items (admin function)
  static async getAllWatchlistItems(): Promise<WatchlistItem[]> {
    try {
      const response = await apiClient.get(this.baseUrl);
      return response.data;
    } catch (error) {
      console.error('Error fetching all watchlist items:', error);
      throw error;
    }
  }
}
